packages/cli/src/lib/oauth/oauthClientServer.ts

```diff
@@ -0,0 +1,102 @@
+import axios, {AxiosRequestConfig} from 'axios';
+import {createServer, IncomingMessage, ServerResponse} from 'http';
+import {URLSearchParams} from 'url';
+import {AuthorizationError, InvalidStateError} from './authorizationError';
+import {AuthorizationServiceConfiguration, ClientConfig} from './oauthConfig';
+
+export class OAuthClientServer {
+  public constructor(
+    private clientConfig: ClientConfig,
+    private authServiceConfig: AuthorizationServiceConfiguration
+  ) {}
+
+  public startServer(port: number, hostname: string, expectedState: string) {
+    return new Promise<{accessToken: string}>((resolve, reject) => {
+      const listener = async (req: IncomingMessage, res: ServerResponse) => {
+        try {
+          const {code, state} = this.parseUrl(req);
+          const {tokenEndpoint} = this.authServiceConfig;
+
+          if (!state || !code) {
+            return;
+          }
+
+          if (state !== expectedState) {
+            throw new InvalidStateError(state, expectedState);
+          }
+
+          const data = this.getTokenQueryString(code);
+          const authRequest = await axios({
+            method: 'post',
+            url: tokenEndpoint,
+            ...this.requestConfig,
+            data,
+          });
+          const accessToken = authRequest.data.access_token;
+
+          res.end('Close your browser to continue');
```

Comment:
Is this needed ? Can we auto-close the browser for the user ?

Reply:
Yes, it is needed. I thought I could return an HTML page containing a script tag to automatically close the browser, but it does not work every time.

Even if the CLI opens the browser tab, it is not allowed to close it once the OAuth flow ends.
Instead, we get the following error when we try to close it `Scripts may close only the windows that were opened by them.`

However, it works as expected when the user is already logged in to the platform.

But even when the CLI is able to auto-close the browser tab, there is a small glitch in the terminal that I have not been able to figure out yet.

All that to say that the auto-closing is currently creating more issues rather than solving the existing one.